dds/DCPS/RTPS/Spdp.cpp

```diff
@@ -965,16 +1014,30 @@ Spdp::data_received(const DataSubmessage& data,
     return;
   }
 
-  handle_participant_data(
-    (data.inlineQos.length() && disposed(data.inlineQos)) ?
-      DCPS::DISPOSE_INSTANCE : DCPS::SAMPLE_DATA,
-    pdata, to_opendds_seqnum(data.writerSN), from, false);
+  const bool from_relay = from == config_->spdp_rtps_relay_address();
+  if (!from_relay && !ip_in_locator_list(from, pdata.participantProxy.metatrafficUnicastLocatorList)) {
+    if (DCPS::DCPS_debug_level) {
+      ACE_DEBUG((LM_DEBUG, ACE_TEXT("(%P|%t) Spdp::data_received - IP not in locator list: %C\n"), DCPS::LogAddr(from).c_str()));
+    }
+    return;
+  }
 
 #ifdef OPENDDS_SECURITY
   if (!is_security_enabled()) {
+    if (!ip_in_AgentInfo(from, plist)) {
```

Comment:
This condition needs to be combined with the previous condition.
Thus, you should have one check if security is enabled that includes the relay, locator list, and agent info and another check for no security that only includes the relay and locator list.

Reply:
The current code seems to do just that. Let's see where it's not what we want.
Three checks: 1. from_relay, 2. in_locator_list, 3. in_AgentInfo (if OPENDDS_SECURITY)

Steps:

1. check from_relay
1.1 Y, check in_AgentInfo (if OPENDDS_SECURITY)
1.2 N, check in_locator_list

2. check in_locator_list
2.1 Y, check in_AgentInfo (if OPENDDS_SECURITY)
2.2 N, return

3. check in_AgentInfo (if OPENDDS_SECURITY)
3.1 Y, process_participant_ice
3.2 N, return

4. if not yet returned, handle_participant_data 

Do you want to check in_AgentInfo first?